pallets/orml-currencies-allowance-ext/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		// RecoverFromErrors { new_status: StatusCode, cleared_errors: Vec<ErrorCode> },
+		UpdateActiveBlock {
+			block_number: T::BlockNumber,
+		},
+		/// (Additional) funds have been approved for transfer to a destination account.
+		ApprovedTransfer {
+			currency_id: CurrencyOf<T>,
+			source: T::AccountId,
+			delegate: T::AccountId,
+			amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Parachain is not running.
+		Unapproved,
+		ParachainNotRunning,
+		CurrencyNotLive,
+	}
+
+	#[pallet::storage]
+	/// Approved balance transfers. Balance is the amount approved for transfer.
+	/// First key is the asset ID, second key is the owner and third key is the delegate.
+	pub type Approvals<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_128Concat, CurrencyOf<T>>,
+			NMapKey<Blake2_128Concat, T::AccountId>, // owner
+			NMapKey<Blake2_128Concat, T::AccountId>, // delegate
+		),
+		BalanceOf<T>,
+	>;
+
+	#[pallet::storage]
+	/// Currencies that can be used to give approval
+	pub(super) type AllowedCurrencies<T: Config> =
+		StorageMap<_, Blake2_128Concat, CurrencyOf<T>, bool>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub allowed_currencies: Vec<CurrencyOf<T>>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { allowed_currencies: vec![] }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for i in &self.allowed_currencies.clone() {
+				AllowedCurrencies::<T>::insert(i, true);
+			}
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	// The pallet's dispatchable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+#[allow(clippy::forget_non_drop, clippy::swap_ptr_to_ref, clippy::forget_ref, clippy::forget_copy)]
+#[cfg_attr(test, mockable)]
+impl<T: Config> Pallet<T> {
+	// Check the amount approved to be spent by an owner to a delegate
+	pub fn allowance(
+		asset: CurrencyOf<T>,
+		owner: &T::AccountId,
+		delegate: &T::AccountId,
+	) -> BalanceOf<T> {
+		Approvals::<T>::get((asset, &owner, &delegate)).unwrap_or_else(Zero::zero)
+	}
+
+	/// Creates an approval from `owner` to spend `amount` of asset `id` tokens by 'delegate'
+	/// while reserving `T::ApprovalDeposit` from owner
+	///
+	/// If an approval already exists, the new amount is added to such existing approval
+	pub fn do_approve_transfer(
+		id: CurrencyOf<T>,
+		owner: &T::AccountId,
+		delegate: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		//TODO ensure for production!!!
+		// ensure!(AllowedCurrencies::<T>::get(id) == Some(true), Error::<T>::CurrencyNotLive);
```

Comment:
What about this comment? Shouldn't we uncomment the ensure statement now?

Reply:
yes. i need to add extrinsic to specify currencies id that we want to support in our chain inside chain extension.